core/modules/widgets/widget.js

```diff
@@ -238,21 +238,64 @@ Widget.prototype.getStateQualifier = function(name) {
 /*
 Compute the current values of the attributes of the widget. Returns a hashmap of the names of the attributes that have changed
 */
+var AttributeTypes = null;
```

Comment:
hmm, ... shouldn't this be part of the widget init mechanism?

Reply:
No, AttributeTypes only need to be loaded once, like widget modules.  Placing the property here is crude, I agree — guidance is welcome.

Reply:
@Jermolene ?

Reply:
Ah, I found a convention to follow:

```javascript
// (in Widget.initialise)
	// Hashmap of the widget classes
	if(!this.widgetClasses) {
		Widget.prototype.widgetClasses = $tw.modules.applyMethods("widget");
	}
```